internal/action/actions.go

```diff
@@ -160,6 +160,52 @@ func (h *BufPane) Center() bool {
 	return true
 }
 
+// CursorToViewTop moves the cursor to the top of the view,
+// offset by scrollmargin unless at the beginning or end of the file
+func (h *BufPane) CursorToViewTop() bool {
+	v := h.GetView()
+	h.Buf.ClearCursors()
+	scrollmargin := int(h.Buf.Settings["scrollmargin"].(float64))
+	if v.StartLine.LessThan(h.Scroll(display.SLoc{0, 0}, scrollmargin)) {
+		scrollmargin = 0
+	}
+	h.Cursor.GotoLoc(h.LocFromVLoc(display.VLoc{
+		SLoc:    h.Scroll(v.StartLine, scrollmargin),
+		VisualX: 0,
+	}))
+	return true
+}
+
+// CursorToViewCenter moves the cursor to the center of the view
+func (h *BufPane) CursorToViewCenter() bool {
+	v := h.GetView()
+	h.Buf.ClearCursors()
+	h.Cursor.GotoLoc(h.LocFromVLoc(display.VLoc{
+		SLoc:    h.Scroll(v.StartLine, h.BufView().Height/2),
+		VisualX: 0,
+	}))
+	return true
+}
+
+// CursorToViewBottom moves the cursor to the bottom of the view,
+// offset by scrollmargin unless at the beginning or end of the file
+func (h *BufPane) CursorToViewBottom() bool {
+	v := h.GetView()
+	h.Buf.ClearCursors()
+	scrollmargin := int(h.Buf.Settings["scrollmargin"].(float64))
+	height := h.BufView().Height
+	bEnd := h.SLocFromLoc(h.Buf.End())
+	lastLine := h.Scroll(v.StartLine, height-1-scrollmargin)
+	if bEnd.Line-lastLine.Line < scrollmargin {
```

Comment:
I'm not seeing the problem. When I test the functionality with wrapped lines, it works as expected. So the code above does the job, but if you have a better solution why not provide it?

Reply:
I'm seeing the problem. Exactly symmetrical to the problem with `CursorToViewTop` that you've just fixed, i.e.: add a very long line as the *last* line in the buffer, slightly scroll the view up so that the end of this last line is outside the view, then trigger `CursorToViewBottom`, and then move the cursor with an arrow key. The view will jump a few lines up due to scrollmargin.

And the solution is also exactly symmetrical to the solution for `CursorToViewTop`.

The idea of `SLoc` is very simple: since the only challenge is that we need to deal not with logical lines in the file but with visual lines on the screen (but otherwise the desired logic is the same as if we had no softwrap), and those visual lines are represented by `SLoc` structures, we simply replace simple arithmetic operations on logical line numbers with corresponding arithmetic operations on `SLoc` structures: `Scroll()` to "add" a number to a `SLoc` (i.e. get the `SLoc` that is N visual lines below the given `SLoc`) and `Diff()` to "subtract" two `SLoc`s (i.e. get the distance between them, in visual lines).

So the solution here is, for example, to replace your

```
	if bEnd.Line-lastLine.Line < scrollmargin {
```

with corresponding

```
	if h.Diff(lastLine, bEnd) < scrollmargin {
```

Only that to make it actually work correctly, need also to fix another bug in your code: replace this:

```
	lastLine := h.Scroll(v.StartLine, height-1-scrollmargin)
```

with this:

```
	lastLine := h.Scroll(v.StartLine, height-1)
```

...And optionally, on top of that, a beautification: replace this:

```
		SLoc:    h.Scroll(v.StartLine, height-1-scrollmargin),
```

with:

```
		SLoc:    h.Scroll(lastLine, -scrollmargin),
```

Reply:
OK, I think I finally got it. There was another bug in the previous commit as well. The condition should be ` == 0` instead of ` < scrollmargin`. I've tested the latest commit with both non-wrapped and wrapped lines and it works well.

Reply:
> There was another bug in the previous commit as well. The condition should be ` == 0` instead of ` < scrollmargin`.

Hmm, no, I think it should be `< scrollmargin`. When the view is almost at the end of the file, if the amount of invisible lines at the end of the file is smaller than `scrollmargin`, then scrollmargin has no effect (try it: move the cursor to the bottom line e.g. by clicking the mouse on the bottom line - the view doesn't shift), yet your latest version of `CursorToViewBottom` needlessly offsets the cursor. (And same with `CursorToViewTop`.)

...Although after more thought, maybe we should rather fix the existing micro's behavior: it seems it would make more sense if `scrollmargin` took effect in such cases. (But even then `CursorToViewTop` and `CursorToViewBottom` would need to be updated as well: the amount of lines they should offset by is then not always `scrollmargin`, it may be smaller than `scrollmargin` if the view is close to the beginning or the end of the file.)

Reply:
And FWIW, `h.Diff(lastLine, bEnd) == 0` is the same thing as `lastLine == bEnd`.

Reply:
Have you tried it with ` < scrollmargin`? Because think about it, the only time we want the cursor to be at the very top/bottom is when the view is already at the very top/bottom. If `scrollmargin` is set to 5 and the view is one line below the top, the diff will be 4, and the cursor will be placed at the top line - which is within the `scrollmargin` and so when the user hits a cursor movement key, the view will jump to have 5 lines of `scrollmargin`.

Reply:
![jump](https://github.com/user-attachments/assets/409f69eb-fba2-4aad-9ceb-58936d2d02f0)

Here's an example.

Reply:
Yes, I meant the diff would be 1, not 4 (now fixed). Did you check the GIF?

Reply:
I only noticed the GIF after I posted my comment. Anyway, I see the GIF shows exactly what I described.

Bottomline is: these annoying jumps (in this corner case when the view is close to the beginning or end of the buffer) are observed also in other scenarios independently of your PR, and it's probably good that your PR avoids those annoying jumps even in this corner case, but we should also fix this corner case for other scenarios (independent of `CursorToViewXXX`) as well.

Reply:
Yes, there are a lot of things we need to fix, but what about this PR? I'd like to get this merged and then open another one to fix the page up/down behavior.

Reply:
I don't think we are in a rush. In the meantime you can clean up the code a little (per my comment: `Diff()` becomes redundant).

Reply:
Fixed.

Reply:
> and then open another one to fix the page up/down behavior.

I was actually going to do that myself (but I barely find free time for anything, so you are also welcome to do it). It seems fairly trivial: basically what I suggested in https://github.com/zyedidia/micro/pull/3506#discussion_r1807894650, plus as people noted in #1808, it would be also useful if `CursorPageUp/Down` scrolled not precisely by height but by height-1 or height-2 to let the reader see some context from the previous page.

Do you think it is more complicated than that?

Reply:
I already have the code ready to go and tested. The way I've done it is to use `scrollmargin` as the number of lines to overlap. I.e., if your view is 110 lines high, and your scrollmargin is 5, Page Down will scroll by 105 lines.

The only question is do you want to replace the existing `CursorPage[Up/Down]` actions, or create new ones.

Reply:
> I already have the code ready to go and tested.

Cool, feel free to submit a PR.

> The way I've done it is to use `scrollmargin` as the number of lines to overlap. I.e., if your view is 110 lines high, and your scrollmargin is 5, Page Down will scroll by 105 lines.

I also thought about that, but it seems like an abuse of `scrollmargin` to me. The documentation says: "`scrollmargin`: margin at which the view starts scrolling when the cursor approaches the edge of the view." so it's a bit different use case. A user may prefer setting `scrollmargin` to 5 lines for scrolling with up/down arrow keys (i.e. for the current use case of `scrollmargin`), but at the same time not want a 5 lines overlap when scrolling with page up/down, that may be too much.

So I'd rather either just hardcode it to e.g. 1 line, or add a separate option for it, e.g. `scrollcontext` or `scrolloverlap`.

> The only question is do you want to replace the existing `CursorPage[Up/Down]` actions, or create new ones.

I'd personally prefer to replace the existing `CursorPage[Up/Down]` actions behavior, since as I said previously, we have no evidence that this behavior is actually preferred by anyone.